onnxruntime/core/providers/cpu/quantization/qlinearconv.cc

```diff
@@ -141,46 +141,97 @@ class QLinearConv : public OpKernel {
-          packed_W_buffer_ = BufferUniquePtr(packed_W, BufferDeleter(alloc));
+    // We need activation and weight zero points for symmetric packing
+    if (!Info().TryGetConstantInput(InputTensors::IN_X_ZERO_POINT, &X_zero_point) || !IsScalarOr1ElementVector(X_zero_point) ||
+        !Info().TryGetConstantInput(InputTensors::IN_W_ZERO_POINT, &W_zero_point) || !IsValidQuantParam(W_zero_point, static_cast<int64_t>(output_channels))) {
+      return false;
+    }
+
+    auto X_zero_point_value = *(X_zero_point->template Data<ActType>());
+    const size_t W_zero_point_size = static_cast<size_t>(W_zero_point->Shape().Size());
+    const auto* W_zero_point_data = W_zero_point->Data<int8_t>();
+    if (!std::all_of(W_zero_point_data, W_zero_point_data + W_zero_point_size, [](int8_t v) { return v == 0; })) {
+      // Symmetric means weight zero point must be zero
+      return false;
+    }
+
+    // Try indirect conv packing
+    size_t packed_size = MlasConvSymPackWSize(group_count, group_input_channels, group_output_channels, kernel_size, std::is_signed<ActType>::value);
+    if (packed_size != 0) {
+      const Tensor* B = nullptr;
+      Info().TryGetConstantInput(8, &B);
+      const auto* Bdata = B != nullptr ? B->template Data<int32_t>() : nullptr;
+
+      column_sums_.resize(output_channels);
+      const int8_t* sdata = (const int8_t*)Wdata;
+      int32_t X_zero_point_fixup = MlasConvSymFixupInputZeroPoint(X_zero_point_value, std::is_signed<ActType>::value);
+      for (size_t oc = 0; oc < output_channels; oc++) {
+        int32_t sum = 0;
+        for (size_t ks = 0; ks < kernel_size * group_input_channels; ks++) {
+          sum += *sdata++;
+        }
+        column_sums_[oc] = (Bdata != nullptr ? Bdata[oc] : 0) - sum * X_zero_point_fixup;
+      }
+
+      auto* packed_W = static_cast<uint8_t*>(alloc->Alloc(packed_size));
+      packed_W_buffer_ = BufferUniquePtr(packed_W, BufferDeleter(alloc));
 
-          MlasConvSymPackW(group_count,
-                           group_input_channels,
-                           group_output_channels,
-                           kernel_size,
-                           reinterpret_cast<const int8_t*>(Wdata),
-                           reinterpret_cast<int8_t*>(packed_W),
-                           packed_size,
-                           std::is_signed<ActType>::value);
+      MlasConvSymPackW(group_count,
+                       group_input_channels,
+                       group_output_channels,
+                       kernel_size,
+                       reinterpret_cast<const int8_t*>(Wdata),
+                       reinterpret_cast<int8_t*>(packed_W),
+                       packed_size,
+                       std::is_signed<ActType>::value);
 
-          is_symmetric_conv_ = true;
+      is_symmetric_conv_ = true;
+      is_W_packed_ = true;
+      return true;
+    }
 
-          is_W_packed_ = true;
-          return true;
+    // Try symmetric GEMM packing
```

Comment:
consolidate this logic with the normal gemm packing?

Reply:
more suggestions?